### ISSUE TITLE
Load balancing + GPU

### ISSUE THREAD

**Issue by marchdf** [AMReX-Codes/amrex#742]
We have the ability for load balancing in PeleC and we currently compute the cost of an MFIter iteration:
```
      if (do_mol_load_balance) {
        wt = (amrex::ParallelDescriptor::second() - wt) / vbox.d_numPts();
        (*cost)[mfi].plus<amrex::RunOn::Device>(wt, vbox);
      }
```
Given the asynchronous nature of the kernel launches, I fear this isn't going to actually give me the correct timings and the load balancing won't work. 

Do you all have a suggestion for getting around this? I looked at the CNS examples GPU and non-GPU but I don't see anything. Maybe I am missing something.

**Comment by atmyers** [comment]
You are correct that this will code will measure CPU execution time, which not relevant when running on the GPU. 

@mrowan137 has been working on this and has an open pull request (PR #616) to amrex that adds the ability to time GPU kernel execution. He can provide more information of the current status of that work.

**Comment by WeiqunZhang** [comment]
You could try `Gpu::streamSynchronize()` before `wt = ...`.  It may not slow down your code because of the stream sync.

**Comment by mrowan137** [comment]
Hi @marchdf , the issue you're describing is precisely what is meant to be addressed in PR #616.  If you're interested in giving it a test for your problem, you could check out this branch: https://github.com/mrowan137/amrex/tree/mrowan/CUPTI_trace
There is an example showing what the instrumented code looks like in `/Tests/GPU/CuptiTest/Exec/CUDA/myfunc.cpp` (look for `BL_PROFILE_VAR_START_CUPTI`, `BL_PROFILE_VAR_STOP_CUPTI`; there are several variations also shown in the example, but this is the basic timing tool.  To use the CUPTI-based timing, you would also need to initialize the CUPTI trace, search for `initCuptiTrace` in `/Tests/GPU/CuptiTest/Source/main.cpp`.  The box timing informations are recorded into a vector `activityRecordUserdata`, and you can get some measure of the net GPU work with function `computeElapsedTimeUserData` (which currently compute average time per kernel, which may not most relevant for you; for something else you could modify the function `computeElapsedTimeUserData` in `/Src/Base/AMReX_CuptiTrace.cpp`), or you could compute something else directly from the activity records stored in `activityRecordUserdata`.  If you end up deciding to give this a try, please don't hesitate to ping me with further questions.

**Comment by marchdf** [comment]
Thanks all for the input. I am going to try `Gpu::streamSynchronize()` and also look into your suggestion @mrowan137.
